### PR TITLE
[DeviceSanitizer] Implement symbolizer for more readable information

### PR DESCRIPTION
**Pull request by zhaomaosu**
SYCLOS Part: https://github.com/intel/llvm/pull/14513

### CODE REVIEW

**Review by yingcong-wu — COMMENTED** [review]
Could you provide a comparison of the output text so that we can see the result difference?

**Comment by zhaomaosu** [comment]
Take the case https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp as example.
Before this change, the output of the case just like:

```
====ERROR: DeviceSanitizer: use-after-free on address 0x118d0f0
READ of size 1 at kernel <typeinfo name for main::{lambda(sycl::_V1::handler&)#1}::operator()(sycl::_V1::handler&) const::MyKernel> LID(0, 0, 0) GID(0, 0, 0)
  #0 main::'lambda'(sycl::_V1::handler&)::operator()(sycl::_V1::handler&) const::'lambda'(sycl::_V1::nd_item<1>)::operator()(sycl::_V1::nd_item<1>) const /export/users/maosuzha/ics_workspace/syclos/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp:16

0x118d0f0 is located inside of Device USM region [0x118d0f0, 0x118d4f0)
allocated here:
  #0 ./use-after-free() [0x4061fe]
  #1 ./use-after-free() [0x404975]
  #2 ./use-after-free() [0x4036e9]
  #3 /lib64/libc.so.6(+0x44e50) [0x7fe60a1a2e50]
  #4 /lib64/libc.so.6(__libc_start_main+0x7c) [0x7fe60a1a2efc]
  #5 ./use-after-free() [0x403595]

released here:
  #0 ./use-after-free() [0x403750]
  #1 /lib64/libc.so.6(+0x44e50) [0x7fe60a1a2e50]
  #2 /lib64/libc.so.6(__libc_start_main+0x7c) [0x7fe60a1a2efc]
  #3 ./use-after-free() [0x403595]
```

And after this change, the output is:

```
====ERROR: DeviceSanitizer: use-after-free on address 0xff7c2c
READ of size 1 at kernel <typeinfo name for main::{lambda(sycl::_V1::handler&)#1}::operator()(sycl::_V1::handler&) const::MyKernel> LID(0, 0, 0) GID(300, 0, 0)
  #0 main::'lambda'(sycl::_V1::handler&)::operator()(sycl::_V1::handler&) const::'lambda'(sycl::_V1::nd_item<1>)::operator()(sycl::_V1::nd_item<1>) const /export/users/maosuzha/ics_workspace/syclos/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp:16

0xff7c2c is located inside of Device USM region [0xff7b00, 0xff7f00)
allocated here:
  #0 in char* sycl::_V1::malloc_device<char>(unsigned long, sycl::_V1::device const&, sycl::_V1::context const&, sycl::_V1::property_list const&, sycl::_V1::detail::code_location const&) /export/users/maosuzha/ics_workspace/syclos/build/bin/../include/sycl/usm.hpp:172:3
  #1 in char* sycl::_V1::malloc_device<char>(unsigned long, sycl::_V1::queue const&, sycl::_V1::property_list const&, sycl::_V1::detail::code_location const&) /export/users/maosuzha/ics_workspace/syclos/build/bin/../include/sycl/usm.hpp:180:10
  #2 in main /export/users/maosuzha/ics_workspace/syclos/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp:10:17
  #3 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7e50)
  #4 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7efc)
  #5 in _start (./use-after-free+0x403595)

released here:
  #0 in main /export/users/maosuzha/ics_workspace/syclos/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp:11:3
  #1 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7e50)
  #2 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7efc)
  #3 in _start (./use-after-free+0x403595)
```

**Comment by zhaomaosu** [comment]
Take the case https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp as a example.

Before this change, the output just like:

```
====ERROR: DeviceSanitizer: use-after-free on address 0x118d0f0
READ of size 1 at kernel <typeinfo name for main::{lambda(sycl::_V1::handler&)#1}::operator()(sycl::_V1::handler&) const::MyKernel> LID(0, 0, 0) GID(0, 0, 0)
  #0 main::'lambda'(sycl::_V1::handler&)::operator()(sycl::_V1::handler&) const::'lambda'(sycl::_V1::nd_item<1>)::operator()(sycl::_V1::nd_item<1>) const .*use-after-free.cpp:16

0x118d0f0 is located inside of Device USM region [0x118d0f0, 0x118d4f0)
allocated here:
  #0 ./use-after-free() [0x4061fe]
  #1 ./use-after-free() [0x404975]
  #2 ./use-after-free() [0x4036e9]
  #3 /lib64/libc.so.6(+0x44e50) [0x7fe60a1a2e50]
  #4 /lib64/libc.so.6(__libc_start_main+0x7c) [0x7fe60a1a2efc]
  #5 ./use-after-free() [0x403595]

released here:
  #0 ./use-after-free() [0x403750]
  #1 /lib64/libc.so.6(+0x44e50) [0x7fe60a1a2e50]
  #2 /lib64/libc.so.6(__libc_start_main+0x7c) [0x7fe60a1a2efc]
  #3 ./use-after-free() [0x403595]
```

After this change, the output just like:

```
====ERROR: DeviceSanitizer: use-after-free on address 0xff7c2c
READ of size 1 at kernel <typeinfo name for main::{lambda(sycl::_V1::handler&)#1}::operator()(sycl::_V1::handler&) const::MyKernel> LID(0, 0, 0) GID(300, 0, 0)
  #0 main::'lambda'(sycl::_V1::handler&)::operator()(sycl::_V1::handler&) const::'lambda'(sycl::_V1::nd_item<1>)::operator()(sycl::_V1::nd_item<1>) const .*use-after-free.cpp:16

0xff7c2c is located inside of Device USM region [0xff7b00, 0xff7f00)
allocated here:
  #0 in char* sycl::_V1::malloc_device<char>(unsigned long, sycl::_V1::device const&, sycl::_V1::context const&, sycl::_V1::property_list const&, sycl::_V1::detail::code_location const&) .*syclos/build/bin/../include/sycl/usm.hpp:172:3
  #1 in char* sycl::_V1::malloc_device<char>(unsigned long, sycl::_V1::queue const&, sycl::_V1::property_list const&, sycl::_V1::detail::code_location const&) .*syclos/build/bin/../include/sycl/usm.hpp:180:10
  #2 in main .*syclos/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp:10:17
  #3 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7e50)
  #4 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7efc)
  #5 in _start (./use-after-free+0x403595)

released here:
  #0 in main .*syclos/sycl/test-e2e/AddressSanitizer/use-after-free/use-after-free.cpp:11:3
  #1 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7e50)
  #2 in __stop___libc_freeres_ptrs (/lib64/libc.so.6+0x7f811dea7efc)
  #3 in _start (./use-after-free+0x403595)
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -37,6 +42,7 @@ option(UR_USE_MSAN "enable MemorySanitizer" OFF)
 option(UR_USE_TSAN "enable ThreadSanitizer" OFF)
 option(UR_ENABLE_TRACING "enable api tracing through xpti" OFF)
 option(UR_ENABLE_SANITIZER "enable device sanitizer" ON)
+option(UR_ENABLE_SYMBOLIZER "enable Symoblizer for Sanitizer" OFF)
```
Comment (AllanZyne) [inline CMakeLists.txt:45]:
```suggestion
option(UR_ENABLE_SYMBOLIZER "enable symbolizer for sanitizer" OFF)
```

File: source/loader/layers/sanitizer/linux/backtrace.cpp [hunk 2]
```diff
@@ -25,8 +67,24 @@ StackTrace GetCurrentBacktrace() {
...
+                if (SrcInfo.file != "??") {
+                    OS << "in " << SrcInfo.function << " " << SrcInfo.file
+                       << ":" << SrcInfo.line << ":" << SrcInfo.column;
+                } else {
+                    OS << "in " << SrcInfo.function << " (" << ModuleName << "+"
+                       << Frames[i] << ")";
+                }
+                Stack.stack.emplace_back(OS.str());
+                continue;
+            }
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/linux/backtrace.cpp:85]: I know I was a bit hasty in implementing this part at first, in fact, the symbolization should be invoked as late as possible.
Can you save as few as possible information when doing backtrace, and symbolize them when really need it?
Thank you!

Comment (AllanZyne) [reply source/loader/layers/sanitizer/linux/backtrace.cpp:85]: I think we can get handle of the module, not file path

Comment (yingcong-wu) [reply source/loader/layers/sanitizer/linux/backtrace.cpp:85]: > I know I was a bit hasty in implementing this part at first, in fact, the symbolization should be invoked as late as possible. Can you save as few as possible information when doing backtrace, and symbolize them when really need it? Thank you!

I agree, the symbolization takes time, so we should only do it when necessary (in error reporting stage I think).

**Comment by yingcong-wu** [comment]
The new error reporting message looks good and more useful, nice improvement!

File: CMakeLists.txt (hunk 1 above)
Comment (zhaomaosu) [reply CMakeLists.txt]: done

File: source/loader/layers/sanitizer/linux/backtrace.cpp (hunk 2 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/linux/backtrace.cpp]: nice suggestions. updated.

**Review by yingcong-wu — APPROVED** [review]
lgtm

**Comment by yingcong-wu** [comment]
Please provide the link to the PR of intel/llvm.

**Review by omarahmed1111 — APPROVED** [review]

**Comment by omarahmed1111** [comment]
@jinge90 Could you review this when possible as this should be ready to merge. Thanks!

**Pull request merged**